### PR TITLE
CA-391651: Fix spike in derived RRD metrics

### PR DESCRIPTION
**Pull request by last-genius**
`xcp-rrdd` used to ignore timestamps of measurement times provided by the RRDD plugins, and always calculated change in all values against a single timestamp captured at the time of reading on the server. This can potentially trigger a case where the time that passed between measurements is not equal to the time that passed between readings (or other measurements), and values would be calculated incorrectly. This can be triggered on a host with 100% CPU load, where `cpu0-P0` derived metric, for example, would often be > 1.0.

This PR reworks `xcp-rrdd` and the plugin infrastructure to calculate values based on the actual length of time passed since the last value for this particular metric was recorded. This required extensive architectural changes, explained in detail in commits themselves - *so best reviewed by commit*.

I've tested this over the weekend, and haven't registered any spikes in metrics on a fully-loaded host. Will run BST+BVT as well, since RRDD is at the center of a lot of other things and is quite fragile.

### CODE REVIEW

**Comment by last-genius** [comment]
NOTE: There's some additional complexity in converting these timestamps to use ptime and mtime for even greater precision, so I've opted to do this separately later on

File: ocaml/libs/xapi-rrd/unix/rrd_unix.ml [hunk 1]
```diff
@@ -30,12 +30,13 @@ let with_out_channel_output fd f =
...
+  with_out_channel_output fd (Rrd.xml_to_output internal rrd)
 
 let json_to_fd rrd fd =
   let payload = Rrd.json_to_string rrd |> Bytes.unsafe_of_string in
   let len = Bytes.length payload in
   Unix.write fd payload 0 len |> ignore
 
-let to_fd ?(json = false) rrd fd =
-  (if json then json_to_fd else xml_to_fd) rrd fd
+let to_fd ?(json = false) ?(internal = true) rrd fd =
```
Comment (psafont) [inline ocaml/libs/xapi-rrd/unix/rrd_unix.ml:41]: I'd prefer the polarity of the parameter be inverted here, that is, that whenever the internal format is needed, it needs to be made explicit

File: ocaml/xcp-rrdd/lib/transport/base/rrd_protocol_v2.ml [hunk 2]
```diff
@@ -75,7 +75,8 @@ module Read = struct
   let datasource_count cs =
     Int32.to_int (Cstruct.BE.get_uint32 cs datasource_count_start)
 
-  let timestamp cs = Cstruct.BE.get_uint64 cs timestamp_start
+  let timestamp cs =
+    Int64.float_of_bits (Cstruct.BE.get_uint64 cs timestamp_start)
```
Comment (psafont) [inline ocaml/xcp-rrdd/lib/transport/base/rrd_protocol_v2.ml:79]: Thinking about this... I don't think we can get away with the breaking change, there are python-based (GFS2) and C-based plugins (not sure which ones they are) :(

Comment (last-genius) [reply ocaml/xcp-rrdd/lib/transport/base/rrd_protocol_v2.ml:79]: After a discussion with the storage team, the timestamp-responsible part is in `rrdd.py` that we control, we'd just need to get a script to restart their metrics collection after update, so this should be do-able

**Comment by last-genius** [comment]
BST+BVT has passed (XenRT run #207011), but there are necessary additions to be made, keeping as a draft. On the other hand, it's sad that at least some of the tests weren't exercising the plugins this must have broken by changing the protocol....

File: ocaml/libs/xapi-rrd/lib/rrd.ml [hunk 3]
```diff
@@ -968,9 +981,7 @@ let xml_to_output rrd output =
     (fun output ->
       tag "version" (data "0003") output ;
       tag "step" (data (Int64.to_string rrd.timestep)) output ;
-      tag "lastupdate"
-        (data (Printf.sprintf "%Ld" (Int64.of_float rrd.last_updated)))
-        output ;
+      tag "lastupdate" (data (Utils.f_to_s rrd.last_updated)) output ;
```
Comment (psafont) [inline ocaml/libs/xapi-rrd/lib/rrd.ml:992]: I don't think this change is explained in the commit message, could you add that this adds decimals to the output format (and this is why the external interface needs to be maintained, which is mentioned in the commit message)

**Comment by lindig** [comment]
I would not worry about backports. We lived with the problematic behavior for so long that that I don't think a risky backport can be justified just to get better stats.

**Review by robhoes — APPROVED** [review]
This is an impressive set of changes. I've reviewed commit-by-commit and every step makes sense (though I haven't checked all code in detail). I assume it is the case that RRD plugins do not need any code change, as long as they are using the updated library (be it OCaml, python or C). There are also no changed external interfaces.

The main upgrade case is that all plugins need to be restarted as soon as xcp-rrdd has restarted to the new version (on a host). Before a restart, xcp-rrdd will interpret int64s as doubles, which means that it won't necessarily fail, but all data would be completely wrong? Then to upgrade, we need to 1) stop xcp-rrdd, 2) install the new xcp-rrdd and libraries, 3) restart all plugins and 4) start xcp-rrdd. Correct?

**Comment by last-genius** [comment]
> The main upgrade case is that all plugins need to be restarted as soon as xcp-rrdd has restarted to the new version (on a host). Before a restart, xcp-rrdd will interpret int64s as doubles, which means that it won't necessarily fail, but all data would be completely wrong? Then to upgrade, we need to 1) stop xcp-rrdd, 2) install the new xcp-rrdd and libraries, 3) restart all plugins and 4) start xcp-rrdd. Correct?

Yes!
Before these changes, xcp-rrdd ignored timestamps provided by the plugins altogether, so if an old xcp-rrdd reads new plugins, nothing should break or change. If a new xcp-rrdd reads old plugins, then yes, the data will be slightly wrong. I've opened a draft PR in xapi.spec that modifies the update guidance as suggested here.

**Comment by robhoes** [comment]
>Before these changes, xcp-rrdd ignored timestamps provided by the plugins altogether, so if an old xcp-rrdd reads new plugins, nothing should break or change.

Ah right, that simplifies the story a little, as we can update and restart plugins first.

**Comment by psafont** [comment]
> Before these changes, xcp-rrdd ignored timestamps provided by the plugins altogether, so if an old xcp-rrdd reads new plugins, nothing should break or change.

Because of this, plugins can be updated before xcp-rrdd is without causing any breakage

**Review by lindig — APPROVED** [review]

**Comment by last-genius** [comment]
I think this should be good to go, see the open xapi.spec PR for how the updates are handled.

**Comment by last-genius** [comment]
xapi.spec PR for restarting the dependent plugins was merged, so merging this as well 🤞

File: ocaml/xcp-rrdd/lib/transport/base/rrd_protocol_v2.ml (hunk 2 above)
Comment (edwintorok) [reply ocaml/xcp-rrdd/lib/transport/base/rrd_protocol_v2.ml:79]: Looking at this later it looks like we actually broke the C plugins completely: https://github.com/xapi-project/rrd-client-lib/pull/15

And storing a double in big-endian is very awkward, I'm not aware of any protocol that would do it that way...

See the linked PR on how complicated it is to actually fill in the right values: you have to get a double in C, take its bits as int64, then convert endianness (without truncating the microseconds).

Would've been a lot easier to just change this to read a double directly, but I guess we're stuck with the awkward protocol now :(

Comment (edwintorok) [reply ocaml/xcp-rrdd/lib/transport/base/rrd_protocol_v2.ml:79]: We could try to change the protocol again and drop the bigendian conversion, but we should first add some (XenRT) tests for this area to check we don't break something again. 

For now I think we'll fix the C plugin to match the protocol expected here.

**Pull request merged**